CRM/Member/Form/Task/PDFLetterCommon.php

```diff
@@ -31,7 +31,16 @@ public static function postProcessMembers(&$form, $membershipIDs, $skipOnHold, $
         $html_message,
         $categories
       );
-    self::createActivities($form, $html_message, $contactIDs, $formValues['subject'], CRM_Utils_Array::value('campaign_id', $formValues));
+
+    // Create activities only if the user has clicked the Download Documents
+    // button (_qf_PDFLetter_upload). Skip activity creation if the user has
+    // clicked the Preview button.
+    $buttonName = $form->controller->getButtonName();
+    $c = $form->controller->container();
+    $isLiveMode = ($buttonName == '_qf_PDFLetter_upload') || isset($c['values']['PDF']['buttons']['_qf_PDFLetter_upload']);
+    if ($isLiveMode) {
+      self::createActivities($form, $html_message, $contactIDs, $formValues['subject'], CRM_Utils_Array::value('campaign_id', $formValues));
+    }
```

Comment:
@andrewpthompson I'd like to see this block of code extracted into a separate function that can be shared between 'CRM_Member_Form_Task_PDFLetterCommon' and 'CRM_Contact_Form_Task_PDFLetterCommon'.

Reply:
@mattwire Thanks for the suggestions Matt. Indeed the same bug is also in contributions.

In trying to make the new function cater for contacts, memberships, contributions and cases it grew a bit. This is because of some differences in the way that `CRM_Contact_Form_Task_PDFLetterCommon::createActivities()` is called by each of the pdf forms:
- The button name is different on the membership form so need to test for 2 different names (I couldn't really see whether or not it's feasible to make the button names consistent)
- For contributions only there is a default subject `ts('Thank you letter')` if none is present in the `$formValues` array
- The `$perContactHtml` parameter of `CRM_Contact_Form_Task_PDFLetterCommon::createActivities()` is used for contributions only
- In `CRM_Contact_Form_Task_PDFLetterCommon::postProcess()` only it makes use of the return value of `CRM_Contact_Form_Task_PDFLetterCommon::createActivities()` (activity IDs array) so we'd need to pass that through.

So I would end up with something like this, which would go in `CRM_Contact_Form_Task_PDFLetterCommon`:
```
  public static function isLiveMode($form, $html_message, $contactIds, $formValues, $perContactHtml = array(), $defaultSubject = NULL) {
    $buttonName = $form->controller->getButtonName();
    $c = $form->controller->container();
    $isLiveMode = ($buttonName == '_qf_PDF_upload') || isset($c['values']['PDF']['buttons']['_qf_PDF_upload']) ||
  ($buttonName == '_qf_PDFLetter_upload') || isset($c['values']['PDF']['buttons']['_qf_PDFLetter_upload']);

    if ($isLiveMode) {
      // Create activities and return an array of the activity IDs
      return self::createActivities($form, $html_message, $contactIds, CRM_Utils_Array::value('subject', $formValues, $defaultSubject), CRM_Utils_Array::value('campaign_id', $formValues), $perContactHtml);
    }
  }
```
I'm thinking this function needs a different name now because isLiveMode implies it would return a value indicating 'not preview mode' but instead it calls `createActivities()` and returns activity IDs. Could I have your thoughts please?

Alternatively instead of creating a new function it could go in the existing `CRM_Contact_Form_Task_PDFLetterCommon::createActivities()`, which is already common to contact/membership/case/contribution, and placed at the top of that function and return prior to actually creating activities if previewing.

Reply:
@andrewpthompson Can you call `isLiveMode()` (or whatever you'd like to call it) from `CRM_Contact_Form_Task_PDFLetterCommon::createActivities()` and return TRUE/FALSE? Then you can return early or continue in `createActivities()` based on that result.
The fact we are using different button names is not great but bringing the button names together into a single function makes it easier to understand and refactor in the future - can you add a comment to indicate where each button name is used?

Reply:
@mattwire Good suggestions, thanks. I've done it that way.